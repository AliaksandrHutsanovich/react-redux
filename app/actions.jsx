var addCategory = function (category) {
    return {
        type: "ADD_CATEGORY",
        category
    }
};

var addSubCategory = function (category, parentCategory) {
    return {
        type: "ADD_SUBCATEGORY",
        category,
        parentCategory
    }
}

var deleteCategory = function(category) {
    return {
        type: "DELETE_CATEGORY",
        category
    }
}

var editCategory = function(newTitleCategory, oldTitleCategory) {
    return {
        type: "EDIT_CATEGORY",
        newTitleCategory,
        oldTitleCategory
    }
}

var addTask = function (task, category) {
    return {
        type: "ADD_TASK",
        category,
        task
    }
}

var editTask = function (oldTask, newTask, oldTitleCategory, newTitleCategory) {
    return {
        type: "EDIT_TASK",
        oldTask,
        newTask,
        oldTitleCategory,
        newTitleCategory
    }
}

var makeVisibleSubCategory = function(subCategory, visible) {
    return {
        type: "MAKE_VISIBLE_SUBCATEGORY",
        subCategory,
        visible
    }
}

var makeUnvisibleSubCategory = function(subCategory, visible) {
    return {
        type: "MAKE_UNVISIBLE_SUBCATEGORY",
        subCategory,
        visible
    }
}

 export default {addCategory, addSubCategory, deleteCategory, editCategory, addTask, editTask, makeVisibleSubCategory, makeUnvisibleSubCategory};