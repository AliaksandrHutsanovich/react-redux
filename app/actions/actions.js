import { createAction } from 'redux-actions';

export const startAddCategoryProcess = createAction('START_ADD_CATEGORY_PROCESS');
export const startAddSubCategoryProcess = createAction('START_ADD_SUB_CATEGORY_PROCESS');
export const startDeleteCategoryProcess = createAction('START_DELETE_CATEGORY_PROCESS');
export const startEditCategoryProcess = createAction('START_EDIT_CATEGORY_PROCESS');
export const startEditTaskStatusProcess = createAction('START_EDIT_TASK_STATUS_PROCESS');
export const startAddTaskProcess = createAction('START_ADD_TASK_PROCESS');
export const startDeleteTaskProcess = createAction('START_DELETE_TASK_PROCESS');
export const startEditTaskProcess = createAction('START_EDIT_TASK_PROCESS');

export const addCategory = createAction('ADD_CATEGORY');
export const addSubCategory = createAction('ADD_SUBCATEGORY');
export const deleteCategory = createAction('DELETE_CATEGORY');
export const editCategory = createAction('EDIT_CATEGORY');

export const editCategoryRevive = createAction('EDIT_CATEGORY_REVIVE');

export const insertCategory = createAction('INSERT_CATEGORY');

export const addTask = createAction('ADD_TASK');
export const editTaskStatus = createAction('EDIT_TASK_STATUS');
export const editTask = createAction('EDIT_TASK');
export const deleteTask = createAction('DELETE_TASK');

export const editTaskRevive = createAction('EDIT_TASK_REVIVE');

export const insertTask = createAction('INSERT_TASK');

export const showDoneTasks = createAction('SHOW_DONE');
export const searchTasks = createAction('SEARCH_TASK');

export const countPercentage = createAction('COUNT_PERCENTAGE');
export const incrementInDone = createAction('INCREMENT');
export const decrementInDone = createAction('DECREMENT');
export const incrementInAll = createAction('INCREMENT_ALL');
export const decrementInAll = createAction('DECREMENT_ALL');

export const startUnDoProcess = createAction('START_UNDO_PROCESS');
export const startReDoProcess = createAction('START_REDO_PROCESS');

export const addToUnDo = createAction('ADD_TO_UNDO');
export const deleteFromUnDo = createAction('DELETE_FROM_UNDO');

export const addToReDo = createAction('ADD_TO_REDO');
export const deleteFromReDo = createAction('DELETE_FROM_REDO');

export const clearReDo = createAction('CLEAR_REDO');
