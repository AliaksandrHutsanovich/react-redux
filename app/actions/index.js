export {
  startAddCategoryProcess,
  startAddSubCategoryProcess,
  startDeleteCategoryProcess,
  startEditCategoryProcess,
  startEditTaskStatusProcess,
  startAddTaskProcess,
  startDeleteTaskProcess,
  startEditTaskProcess,
  addCategory,
  addSubCategory,
  deleteCategory,
  editCategory,
  editCategoryRevive,
  insertCategory,
  addTask,
  editTaskStatus,
  editTask,
  deleteTask,
  editTaskRevive,
  insertTask,
  showDoneTasks,
  searchTasks,
  countPercentage,
  incrementInDone,
  decrementInDone,
  incrementInAll,
  decrementInAll,
  startUnDoProcess,
  startReDoProcess,
  addToUnDo,
  deleteFromUnDo,
  addToReDo,
  deleteFromReDo,
  clearReDo,
  switchContentDisplay,
} from './actions';
