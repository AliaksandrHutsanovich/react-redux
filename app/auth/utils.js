import { fieldTitles } from './constants';

const validatorByField = {
  [fieldTitles.confirm]: (getFieldValue) => (rule, value, callback) => {
    if (value && value !== getFieldValue('password')) {
      callback('Two passwords that you enter is inconsistent!');
    } else {
      callback();
    }
  },
  [fieldTitles.password]: (validateFields, confirmDirty) => (rule, value, callback) => {
    if (value && confirmDirty) {
      validateFields(['confirm'], { force: true });
    }
    callback();
  },
};

export default validatorByField;
