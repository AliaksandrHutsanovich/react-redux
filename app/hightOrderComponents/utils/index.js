export {
  getPathParam,
  getSavingPath,
  caseClickHandlers,
  kindsOfAddActions,
  typesCategoryOperation,
  formNameByPlaceholder,
  getPathByOperation,
} from './utils';
