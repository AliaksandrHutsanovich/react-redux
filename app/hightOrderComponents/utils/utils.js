import OPERATION_TITLES from '../../constants';
import {
  incrementInAll,
  startAddCategoryProcess,
  startAddTaskProcess,
  startAddSubCategoryProcess,
  startEditCategoryProcess,
  startDeleteCategoryProcess,
} from '../../actions';

export const getPathParam = (path) => ({
  savingPath: path.split('-'),
  param: path.split('-')[path.split('-').length - 1],
});

export const getSavingPath = (path) => {
  const { savingPath } = getPathParam(path);
  savingPath.pop();
  return savingPath;
};

export const caseClickHandlers = {
  [OPERATION_TITLES.ADD_NEW_CATEGORY]: (dispatch, title, addAction) => {
    dispatch(addAction({ title }));
  },
  [OPERATION_TITLES.ADD_NEW_TASK]: (dispatch, title, addAction, url) => {
    dispatch(incrementInAll());
    dispatch(addAction({ path: (url.replace('/', '') + '-tasks').split('-'), title }));
  },
};

export const kindsOfAddActions = {
  [OPERATION_TITLES.ADD_NEW_CATEGORY]: startAddCategoryProcess,
  [OPERATION_TITLES.ADD_NEW_TASK]: startAddTaskProcess,
};

export const getPathByOperation = {
  [OPERATION_TITLES.ADD_NEW_SUBCATEGORY]: (path) => (path + '-subCategories').split('-'),
  [OPERATION_TITLES.EDIT_CATEGORY]: (path) => getSavingPath(path),
};

export const typesCategoryOperation = {
  [OPERATION_TITLES.ADD_NEW_SUBCATEGORY]: (path, dispatch, value) => {
    dispatch(startAddSubCategoryProcess({
      path: (path + '-subCategories').split('-'),
      title: value,
    }));
  },
  [OPERATION_TITLES.EDIT_CATEGORY]: (path, dispatch, value, title) => {
    dispatch(startEditCategoryProcess({
      path: getSavingPath(path),
      pathParam: getPathParam(path).param,
      title: value || title,
    }));
  },
  [OPERATION_TITLES.DELETE_CATEGORY]: (path, dispatch) => {
    dispatch(startDeleteCategoryProcess({
      path: getSavingPath(path),
      pathParam: getPathParam(path).param,
    }));
  },
};

export const formNameByPlaceholder = {
  [OPERATION_TITLES.ADD_NEW_CATEGORY]: 'addCategoryForm',
  [OPERATION_TITLES.ADD_NEW_TASK]: 'addTaskForm',
};
