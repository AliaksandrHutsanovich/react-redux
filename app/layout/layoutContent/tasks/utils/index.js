export { default, getOldPathParams } from './utils';
