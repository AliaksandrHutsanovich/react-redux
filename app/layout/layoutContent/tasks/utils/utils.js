import {
  incrementInDone,
  decrementInDone,
} from '../../../../actions';

const changeInDone = (
  isFinishedTask,
  dispatch,
) => (isFinishedTask ? dispatch(incrementInDone()) : dispatch(decrementInDone()));

const getStateValue = (isFinishedTask, isStatusChanged, dispatch) => {
  if (isStatusChanged) {
    changeInDone(isFinishedTask, dispatch);
  }
  return isFinishedTask;
};

export const getOldPathParams = (oldPath) => ({
  oldPath: oldPath.filter((x, index, path) => index < path.length - 1),
  oldPathParam: oldPath[oldPath.length - 1],
});

export default getStateValue;
