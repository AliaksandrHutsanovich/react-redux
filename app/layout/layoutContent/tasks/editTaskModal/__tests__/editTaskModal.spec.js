import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Controller } from 'react-hook-form';
import { Map } from 'immutable';
import { Form, Modal } from 'antd';
import { act } from 'react-dom/test-utils';
import { initialState as store } from '../../../../../reducers';
import EditTaskModal, { Categories, EditTaskModal as WithoutWrapper } from '../editTaskModal';
import { getOldPathParams } from '../../utils';
import { startEditTaskProcess } from '../../../../../actions';

const mockStore = configureStore();
const initialState = Map().merge(store);
const oldPath = '/task';
const taskTitle = 'Buy a new stearing wheal';
const description = 'This task is very important';

const newTestPath = 'categories-0';
const newTasksStatus = false;
const newDescription = 'This task is not done';
const newTaskTitle = 'Change the bought stearing wheal';

jest.mock('../../../../../actions/actions', () => ({
  startEditTaskProcess: jest.fn(),
  clearReDo: () => {},
  decrementInDone: () => {},
}));
jest.mock('../../../../../sagas/execute', () => () => false);

describe('Unit test of edit task modal', () => {
  const dataStore = mockStore({ actionReducers: initialState });
  dataStore.dispatch = jest.fn();

  const Component = shallow(
    <EditTaskModal
      taskTitle={taskTitle}
      description={description}
      isFinished
      oldPath={oldPath}
      handleOk={() => {}}
      visible
      handleCancel={() => {}}
      store={dataStore}
      location={{}}
    />,
  );

  it('Full render test of edit task modal', () => {
    expect(Component.shallow()).toMatchSnapshot();
  });

  it('Category should be clickable', async () => {
    const fragment = shallow(
      <WithoutWrapper
        taskTitle={taskTitle}
        description={description}
        isFinished
        oldPath={oldPath}
        handleOk={() => {}}
        visible
        handleCancel={() => {}}
        store={dataStore}
        dispatch={() => {}}
        location={{}}
      />,
    );
    // set a new tast path
    fragment.find(Categories).prop('onSelectCategory')([newTestPath]);
    const checkbox = fragment.find(Controller).at(1).dive();
    const input = fragment.find(Controller).at(0).dive();
    const textarea = fragment.find(Controller).at(2).dive();
    // change task status
    checkbox.simulate(
      'change',
      {
        target: {
          checked: true,
        },
      },
    );

    checkbox.simulate(
      'change',
      {
        target: {
          checked: newTasksStatus,
        },
      },
    );

    // change task title
    input.simulate(
      'change',
      {
        target: {
          value: newTaskTitle,
        },
      },
    );

    // change task description
    textarea.simulate(
      'change',
      {
        target: {
          value: newDescription,
        },
      },
    );

    const oldPathParams = getOldPathParams(oldPath.split('-'));
    const newPath = (`${newTestPath}-tasks`).split('-');
    const newPathParam = '';
    await fragment.find(Form).prop('onSubmit')();

    expect(startEditTaskProcess).toHaveBeenCalledWith({
      newPath,
      oldPath: oldPathParams.oldPath,
      oldPathParam: oldPathParams.oldPathParam,
      newPathParam,
      title: newTaskTitle,
      description: newDescription,
      isFinished: newTasksStatus,
      location: {},
    });
  });

  it('dialog should be closable', async () => {
    const testFun = jest.fn();
    const fragment = shallow(
      <WithoutWrapper
        taskTitle={taskTitle}
        description={description}
        isFinished
        oldPath={oldPath}
        handleOk={() => {}}
        visible
        handleCancel={testFun}
        store={dataStore}
        dispatch={() => {}}
        location={{}}
      />,
    );

    await act(async () => {
      fragment.find(Modal).prop('onCancel')();
    });
    expect(testFun).toHaveBeenCalled();
  });
});
