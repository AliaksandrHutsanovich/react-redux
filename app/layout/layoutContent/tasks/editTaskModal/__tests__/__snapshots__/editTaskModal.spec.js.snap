// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test of edit task modal Full render test of edit task modal 1`] = `
<Modal
  confirmLoading={false}
  maskTransitionName="fade"
  okButtonProps={
    Object {
      "form": "editTaskForm-Buy a new stearing wheal",
      "htmlType": "submit",
    }
  }
  okText="Save changes"
  okType="primary"
  onCancel={[Function]}
  title="Edit task"
  transitionName="zoom"
  visible={true}
  width="900px"
>
  <Form
    autoComplete="off"
    colon={true}
    data-testid="edit-task-form"
    hideRequiredMark={false}
    id="editTaskForm-Buy a new stearing wheal"
    layout="horizontal"
    onSubmit={[Function]}
  >
    <Row
      gutter={0}
    >
      <Col
        span={12}
      >
        <Connect(Component)
          onSelectCategory={[Function]}
        />
      </Col>
      <Col
        span={12}
      >
        <Row
          gutter={0}
        >
          <Controller
            as={
              <Input
                placeholder="Input task title"
                type="text"
              />
            }
            control={
              Object {
                "defaultValuesRef": Object {
                  "current": Object {},
                },
                "errorsRef": Object {
                  "current": Object {},
                },
                "fieldArrayNamesRef": Object {
                  "current": Set {},
                },
                "fieldsRef": Object {
                  "current": Object {},
                },
                "formState": Object {
                  "dirty": false,
                  "dirtyFields": Set {},
                  "isSubmitted": false,
                  "isSubmitting": false,
                  "isValid": false,
                  "submitCount": 0,
                  "touched": Object {},
                },
                "getValues": [Function],
                "isDirtyRef": Object {
                  "current": false,
                },
                "mode": Object {
                  "isOnBlur": false,
                  "isOnSubmit": true,
                },
                "reValidateMode": Object {
                  "isReValidateOnBlur": false,
                  "isReValidateOnSubmit": false,
                },
                "readFormStateRef": Object {
                  "current": Object {
                    "dirty": false,
                    "dirtyFields": false,
                    "isSubmitted": true,
                    "isSubmitting": false,
                    "isValid": false,
                    "submitCount": false,
                    "touched": false,
                  },
                },
                "register": [Function],
                "removeFieldEventListener": [Function],
                "resetFieldArrayFunctionRef": Object {
                  "current": Object {},
                },
                "setValue": [Function],
                "touchedFieldsRef": Object {
                  "current": Object {},
                },
                "triggerValidation": [Function],
                "unregister": [Function],
                "watchFieldArrayRef": Object {
                  "current": Object {},
                },
              }
            }
            defaultValue="Buy a new stearing wheal"
            name="title"
            rules={
              Object {
                "required": "This is required",
              }
            }
          />
        </Row>
        <Row
          gutter={0}
        >
          <Controller
            as={
              <Checkbox
                indeterminate={false}
              >
                Done
              </Checkbox>
            }
            control={
              Object {
                "defaultValuesRef": Object {
                  "current": Object {},
                },
                "errorsRef": Object {
                  "current": Object {},
                },
                "fieldArrayNamesRef": Object {
                  "current": Set {},
                },
                "fieldsRef": Object {
                  "current": Object {},
                },
                "formState": Object {
                  "dirty": false,
                  "dirtyFields": Set {},
                  "isSubmitted": false,
                  "isSubmitting": false,
                  "isValid": false,
                  "submitCount": 0,
                  "touched": Object {},
                },
                "getValues": [Function],
                "isDirtyRef": Object {
                  "current": false,
                },
                "mode": Object {
                  "isOnBlur": false,
                  "isOnSubmit": true,
                },
                "reValidateMode": Object {
                  "isReValidateOnBlur": false,
                  "isReValidateOnSubmit": false,
                },
                "readFormStateRef": Object {
                  "current": Object {
                    "dirty": false,
                    "dirtyFields": false,
                    "isSubmitted": true,
                    "isSubmitting": false,
                    "isValid": false,
                    "submitCount": false,
                    "touched": false,
                  },
                },
                "register": [Function],
                "removeFieldEventListener": [Function],
                "resetFieldArrayFunctionRef": Object {
                  "current": Object {},
                },
                "setValue": [Function],
                "touchedFieldsRef": Object {
                  "current": Object {},
                },
                "triggerValidation": [Function],
                "unregister": [Function],
                "watchFieldArrayRef": Object {
                  "current": Object {},
                },
              }
            }
            defaultValue={true}
            name="isFinished"
            valueName="checked"
          />
        </Row>
        <Row
          gutter={0}
        >
          <Controller
            as={
              <TextArea
                placeholder="Input type description"
                rows={4}
              />
            }
            control={
              Object {
                "defaultValuesRef": Object {
                  "current": Object {},
                },
                "errorsRef": Object {
                  "current": Object {},
                },
                "fieldArrayNamesRef": Object {
                  "current": Set {},
                },
                "fieldsRef": Object {
                  "current": Object {},
                },
                "formState": Object {
                  "dirty": false,
                  "dirtyFields": Set {},
                  "isSubmitted": false,
                  "isSubmitting": false,
                  "isValid": false,
                  "submitCount": 0,
                  "touched": Object {},
                },
                "getValues": [Function],
                "isDirtyRef": Object {
                  "current": false,
                },
                "mode": Object {
                  "isOnBlur": false,
                  "isOnSubmit": true,
                },
                "reValidateMode": Object {
                  "isReValidateOnBlur": false,
                  "isReValidateOnSubmit": false,
                },
                "readFormStateRef": Object {
                  "current": Object {
                    "dirty": false,
                    "dirtyFields": false,
                    "isSubmitted": true,
                    "isSubmitting": false,
                    "isValid": false,
                    "submitCount": false,
                    "touched": false,
                  },
                },
                "register": [Function],
                "removeFieldEventListener": [Function],
                "resetFieldArrayFunctionRef": Object {
                  "current": Object {},
                },
                "setValue": [Function],
                "touchedFieldsRef": Object {
                  "current": Object {},
                },
                "triggerValidation": [Function],
                "unregister": [Function],
                "watchFieldArrayRef": Object {
                  "current": Object {},
                },
              }
            }
            defaultValue="This task is very important"
            name="description"
          />
        </Row>
      </Col>
    </Row>
  </Form>
</Modal>
`;
